frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/services/JwtService.java

```diff
@@ -0,0 +1,94 @@
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.Payload;
+import com.nimbusds.jose.crypto.RSAEncrypter;
+import com.nimbusds.jose.crypto.impl.ECDSA;
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jwt.EncryptedJWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.kms.model.KmsException;
+import software.amazon.awssdk.services.kms.model.SignRequest;
+import software.amazon.awssdk.services.kms.model.SignResponse;
+import software.amazon.awssdk.services.kms.model.SigningAlgorithmSpec;
+import uk.gov.di.authentication.shared.services.KmsConnectionService;
+
+import java.nio.charset.StandardCharsets;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+
+import static com.nimbusds.jose.crypto.impl.ECDSA.getSignatureByteArrayLength;
+
+public class JwtService {
+    private static final Logger LOG = LogManager.getLogger(JwtService.class);
+    private final KmsConnectionService kmsConnectionService;
+
+    public JwtService(KmsConnectionService kmsConnectionService) {
+        this.kmsConnectionService = kmsConnectionService;
+    }
+
+    public SignedJWT signJWT(
+            JWSAlgorithm signingAlgorithm, JWTClaimsSet jwtClaims, String keyAlias) {
+        LOG.info("Creating signed JWT");
+        JWSHeader jwsHeader = new JWSHeader(signingAlgorithm);
+        Base64URL encodedHeader = jwsHeader.toBase64URL();
+        Base64URL encodedClaims = Base64URL.encode(jwtClaims.toString());
+        String message = encodedHeader + "." + encodedClaims;
+        SignRequest signRequest =
+                SignRequest.builder()
+                        .message(SdkBytes.fromByteArray(message.getBytes(StandardCharsets.UTF_8)))
+                        .keyId(keyAlias)
+                        .signingAlgorithm(SigningAlgorithmSpec.ECDSA_SHA_256)
+                        .build();
+        try {
+            LOG.info("Signing JWT");
+            SignResponse signResult = kmsConnectionService.sign(signRequest);
+            LOG.info("JWT has been signed successfully");
+            String signature =
+                    Base64URL.encode(
+                                    ECDSA.transcodeSignatureToConcat(
+                                            signResult.signature().asByteArray(),
+                                            getSignatureByteArrayLength(signingAlgorithm)))
+                            .toString();
+            return SignedJWT.parse(message + "." + signature);
+        } catch (ParseException | JOSEException e) {
+            LOG.error("Failed to sign JWT", e);
+            throw new RuntimeException(e);
+        } catch (KmsException e) {
+            LOG.error("KMS error when signing JWT", e);
+            throw new RuntimeException(e);
+        }
```

Comment:
Yep happy to switch these around. Just for my sake, is this because KMSException catch block is more specific?

Reply:
It's just because the order in which the exceptions may be thrown is the other way around - KMSException is thrown by the `kmsConnectionService` (line 55), and the JOSEException and ParseException are thrown by `ECDSA.transcodeSignatureToConcat` (line 59) and `SignedJWT.parse` (line 63).

You can totally ignore this though, I'm just being fussy 😆.

Reply:
No worries, that makes sense! I have swapped them around